docs/mem-management/man-vs-auto-and-gc.md

```diff
@@ -0,0 +1,132 @@
+
+```
+The dynamic variables that are used in D programs live on memory blocks that are owned by the garbage collector (GC). When the lifetime of a variable ends (i.e. it's no longer being used), that variable is subject to being finalized according to an algorithm that is executed by the GC. If nothing else needs the memory location containing the variable, the memory may be reclaimed to be used for other variables. This algorithm is called garbage collection and an execution of the algorithm is called a garbage collection cycle.
+```
+
+# Memory management in D
+
+The [D language specification](https://dlang.org/spec/garbage.html) offers a good insight into how the D's GC works, here we will present only the most important aspects of it. D makes extensive use of GC, even if it might not be obvious at a first glance. To make efficient use of it, we must understand when and how the GC does it's job. Exceptions, strings, dynamic and associative arrays are some of the built-in types that make use of GC-allocated memory. Let's look at an example, using the append operator:
+
+```d
+void main() {
+    int[] ints;
+    foreach(i; 0..100) {
+        ints ~= i;
+    }
+}
+```
+
+Appending elements to a dynamic array in this manner will trigger GC allocations when the capacity of the array is insufficient. You can see this by compiling the above code with the option ```-vgc```. Even better, we can use the [capacity](https://dlang.org/phobos/object.html#.capacity) property of the array to inspect the capacity at any time.
+
+```d
+void main() {
+    import std.stdio : writef;
+    int[] ints;
+    size_t before, after, num;
+    writef("%s", ints.capacity);
+    foreach(i; 0..100) {
+        before = ints.capacity;
+        ints ~= i;
+        after = ints.capacity;
+        if(before != after) {
+            writef(" -> %s", after);
+            num++;
+        }
+    }
+    writef("\nNumber of allocations: %s", num);
+}
+```
+
+Now we can see exactly when the allocations take place. The output of the above snippet may look like this:
+```
+0 -> 3 -> 7 -> 11 -> 15 -> 23 -> 31 -> 43 -> 63 -> 91 -> 127
+Number of allocations: 10
+```
+
+If we know exactly how much memory we need, we can preallocate it, triggering a single allocation:
+
+```d
+int[] ints = new int[](100);
+```
+
+# Using the Garbage Collector explicitly
+
+To reclaim memory, the GC keeps track of which blocks of memory are still accessible and which are not. The GC starts from locations known as program roots (implicitly, those are: the stacks of every thread, all global and thread-local variables; you can add locations for the GC to scan for references to memory blocks with ```GC.addRoot``` or ```GC.addRange```) and looks for pointers to memory blocks. The GC will tag reachable blocks as still in use and unreachable blocks will be reclaimed. Because the GC allocated them in the first place, they are still accessible to the GC, so the GC can call their respective destructors.
```

Comment:
Why and when would a user use `addRoot` and `addRange`? Add samples and more details.